activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  module ConnectionAdapters
+    class LazyTransactionProxy < SimpleDelegator # :nodoc:
+      class StatementProxy < SimpleDelegator # :nodoc:
+        def initialize(statement, connection)
+          __setobj__(statement)
+          @connection = connection
+        end
+
+        def method_missing(*)
```

Comment:
This needs to implement `respond_to_missing?` as well

Reply:
`Delegator` already provides an implementation of `respond_to_missing?`:
https://github.com/ruby/ruby/blob/v2_4_4/lib/delegate.rb#L95-L104

Since we call `super` in `method_missing`, we should do the same for `respond_to_missing?`:

```ruby
def respond_to_missing?(*)
  super
end
```

I've added this to both classes, if only to document that using `Delegator`'s behaviour unmodified is intentional.

Reply:
The contract between the two methods is that `respond_to_missing?` will return true iff called with a method that `method_missing` would dispatch. Usually this means making equivalent changes to both methods to keep them in sync, as you've noted. In this case however, calling `materialize_transactions` in `method_missing` doesn't affect which methods are dispatched; it's purely a side effect, and so doesn't require an equivalent change in `respond_to_missing?`.

Another way of looking at it: we need to call `materialize_transactions` in `method_missing` because we don't know whether the call to `super` will touch the socket. We don't need to call it in `respond_to_missing?` because we know that calling `super` won't touch the socket.